pkg/gatewayserver/gatewayserver.go

```diff
@@ -713,11 +713,19 @@ func (gs *GatewayServer) handleLocationUpdates(conn connectionEntry) {
 			status, _, ok := conn.StatusStats()
 			locations := status.AntennaLocations
 			antennas := conn.Gateway().Antennas
-			if ok && len(locations) > 0 && len(antennas) > 0 {
-				// TODO: Handle multiple antenna locations (https://github.com/TheThingsNetwork/lorawan-stack/issues/2006).
-				locations[0].Source = ttnpb.SOURCE_GPS
-				antennas[0].Location = *locations[0]
 
+			// TODO: Handle multiple antenna locations (https://github.com/TheThingsNetwork/lorawan-stack/issues/2006).
+			if ok && len(locations) > 0 {
+				if len(antennas) == 0 {
+					// Add an antenna if none are present
+					conn.Gateway().Antennas = []ttnpb.GatewayAntenna{{}}
+					antennas = conn.Gateway().Antennas
+				}
+				locations[0].Source = ttnpb.SOURCE_REGISTRY
+				if antennas[0].Location.Equal(*locations[0]) {
+					break
+				}
+				antennas[0].Location = *locations[0]
```

Comment:
So now the location is set to registry, but these are GPS values, right?

Reply:
This is actually because the location from the registry never stores location source. https://github.com/TheThingsNetwork/lorawan-stack/blob/6a873f552411bb41ecffb91e46041dbc6982b46c/pkg/identityserver/store/types.go#L124-L131

So using SOURCE_GPS as a source not only has no effect, it will also cause the equality check below to fail because of the difference in location sources.

Fixing that requires a database migration (adding a `Source` to `GatewayAntenna` ORM model. Creating a pull request for that.

Reply:
Reverted change to ttnpb.SOURCE_GPS.

Reply:
Why do we depend on what IS stores exactly? Isn't this just about checking if the location doesn't change at the GS level, before it updates it in IS?

Reply:
The change is checked at the GS level. But location equality check also includes the location source, and that is always hard-coded to `SOURCE_REGISTRY` for gateway locations from the IS at the moment. This only really effects the first status message received (because the initial location on GS is the one retrieved from IS). Discussion is in #2245, not sure if we really care about it.